### PR TITLE
Fix all the typing omissions and errors in libCharon/filetypes. This …

### PR DESCRIPTION
**Pull request by daid**
…also fixes 1 bug in the json metadata reader, which was trying to parse binary data as string and failing because of that.

### CODE REVIEW

**Comment by daid** [comment]
Fine with me, but I have zero Docker experience, so more likely to do something wrong then right on 
that area. Also, there are still a lot of mypy error after this patch, as I only fixed the library. So not sure how to set that up to warn us of mypy issues without getting crazy.
```
Charon\Service\FileService.py:1: error: Cannot find module named 'dbus'
Charon\Service\FileService.py:4: error: Cannot find module named 'RequestQueue'
Charon\Client\DBusInterface.py:12: error: No library stub file for module 'PyQt5.QtCore'
Charon\Client\DBusInterface.py:12: note: (Stub files are from https://github.com/python/typeshed)
Charon\Client\DBusInterface.py:13: error: No library stub file for module 'PyQt5.QtDBus'
Charon\Client\DBusInterface.py:21: error: Cannot find module named 'dbus'
Charon\Client\DBusInterface.py:21: note: (Perhaps setting MYPYPATH or using the "--ignore-missing-imports" flag would help)
Charon\Client\DBusInterface.py:22: error: Cannot find module named 'dbus.mainloop'
Charon\Client\DBusInterface.py:22: error: Cannot find module named 'dbus.mainloop.glib'
Charon\Client\DBusInterface.py:23: error: No library stub file for module 'gi.repository'
Charon\Client\DBusInterface.py:68: error: "None" has no attribute "call"
Charon\Client\DBusInterface.py:74: error: Incompatible return value type (got "None", expected "bool")
Charon\Client\DBusInterface.py:77: error: "None" has no attribute "call_blocking"
Charon\Client\DBusInterface.py:94: error: Missing return statement
Charon\Client\DBusInterface.py:100: error: "None" has no attribute "asyncCall"
Charon\Client\DBusInterface.py:102: error: "None" has no attribute "call_async"
Charon\Client\DBusInterface.py:120: error: "None" has no attribute "addConnection"
Charon\Client\DBusInterface.py:122: error: "None" has no attribute "add_signal_receiver"
Charon\Client\DBusInterface.py:141: error: "None" has no attribute "removeConnection"
Charon\Client\DBusInterface.py:143: error: "None" has no attribute "remove_signal_receiver"
Charon\filetypes\UltimakerFormatPackage.py:309: error: No library stub file for module 'PyQt5.QtGui'
Charon\filetypes\UltimakerFormatPackage.py:310: error: No library stub file for module 'PyQt5.QtCore'
Charon\Service\main.py:5: error: Cannot find module named 'dbus'
Charon\Service\main.py:5: error: Cannot find module named 'dbus.service'
Charon\Service\main.py:6: error: Cannot find module named 'dbus.mainloop'
Charon\Service\main.py:6: error: Cannot find module named 'dbus.mainloop.glib'
Charon\Service\main.py:7: error: No library stub file for module 'gi.repository'
Charon\Service\main.py:21: error: Incompatible types in assignment (expression has type "str", target has type "int")
Charon\Service\main.py:22: error: Argument 1 to "basicConfig" has incompatible type "**Dict[str, int]"; expected "str"
Charon\Service\main.py:22: error: Argument 1 to "basicConfig" has incompatible type "**Dict[str, int]"; expected "IO[str]"
Charon\Service\main.py:22: error: Argument 1 to "basicConfig" has incompatible type "**Dict[str, int]"; expected "Iterable[Handler]"
Charon\Client\Request.py:27: error: The return type of "__init__" must be None
Charon\Client\Request.py:33: error: Need type annotation for '__data'
Charon\Client\Request.py:92: error: Incompatible types in assignment (expression has type "Optional[Callable[[Request, Dict[str, Any]], None]]", variable has type "None")
Charon\Client\Request.py:93: error: Incompatible types in assignment (expression has type "Optional[Callable[[Request], None]]", variable has type "None")
Charon\Client\Request.py:94: error: Incompatible types in assignment (expression has type "Optional[Callable[[Request, str], None]]", variable has type "None")
Charon\Service\RequestQueue.py:4: error: Cannot find module named 'dbus'
Charon\Service\RequestQueue.py:7: error: Cannot find module named 'FileService'
Charon\Service\RequestQueue.py:25: error: The return type of "__init__" must be None
Charon\Client\test_qt.py:3: error: No library stub file for module 'PyQt5.QtCore'
Charon\Client\test_glib.py:3: error: Cannot find module named 'dbus'
Charon\Client\test_glib.py:4: error: Cannot find module named 'dbus.mainloop'
Charon\Client\test_glib.py:4: error: Cannot find module named 'dbus.mainloop.glib'
Charon\Client\test_glib.py:5: error: No library stub file for module 'gi.repository'
Charon\Client\test_glib.py:20: error: Too many positional arguments for "setCallbacks" of "Request"
```

File: Charon/filetypes/UltimakerFormatPackage.py [hunk 1]
```diff
@@ -61,13 +61,13 @@ def openStream(self, stream: BufferedIOBase, mime: str = "application/x-ufp", mo
...
     def close(self) -> None:
-        if not self.stream:
+        if not self.stream or not self.zipfile:
             raise ValueError("This file is already closed.")
         self.flush()
         self.zipfile.close()
 
     def flush(self) -> None:
-        if not self.stream:
+        if not self.stream or not self.zipfile:
```
Comment (nallath) [inline Charon/filetypes/UltimakerFormatPackage.py]: Nitpicking; is it intentional that you use not self.zipfile and self.zipfile is None? Is this intentional? If not, i think it's better to be consistent (and use self.zipfile is None)

File: Charon/filetypes/UltimakerFormatPackage.py [hunk 2]
```diff
@@ -257,6 +259,8 @@ def addRelation(self, virtual_path: str, relation_type: str, origin: str = "") -
     #   \return ``True`` if it exists as a normal resource, or ``False`` if it
     #   doesn't.
     def _resource_exists(self, virtual_path: str) -> bool:
```
Comment (nallath) [inline Charon/filetypes/UltimakerFormatPackage.py]: You didn't write it it, but should be _resourceExists

File: Charon/filetypes/UltimakerFormatPackage.py [hunk 3]
```diff
@@ -410,7 +424,7 @@ def _readMetadata(self) -> None:
                 if metadata_file not in self.zipfile.namelist(): #The metadata file is unknown to us.
                     continue
 
-                metadata = json.load(self.zipfile.open(metadata_file))
+                metadata = json.loads(self.zipfile.open(metadata_file).read().decode('utf-8'))
```
Comment (nallath) [inline Charon/filetypes/UltimakerFormatPackage.py]: Codestyle; should be "utf-8"

File: Charon/filetypes/UltimakerFormatPackage.py (hunk 1 above)
Comment (daid) [reply Charon/filetypes/UltimakerFormatPackage.py]: Actually, these should be just asserts, if stream is not set, then zipfile is also not set. If stream is available and zipfile is not, that indicates an code malfunction.

File: Charon/filetypes/UltimakerFormatPackage.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daid) [reply Charon/filetypes/UltimakerFormatPackage.py]: Fine with me, but what to do when this execption happens, raise an UFPError, to indicate it's not a valid file? Or ignore this json file and log the problem?

File: Charon/filetypes/UltimakerFormatPackage.py (hunk 2 above)
Comment (daid) [reply Charon/filetypes/UltimakerFormatPackage.py]: All comments are fine, trying to cleanup the code a bit more, remember ;-)

**Comment by daid** [comment]
@ChrisTerBeke as our "Docker expert", is there any way to ensure our Docker IC has python 3.4? I think it's best to run the tests on python 3.4 as the firmware is still stuck on that version.

File: Charon/filetypes/UltimakerFormatPackage.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daid) [reply Charon/filetypes/UltimakerFormatPackage.py]: Note, so I did not fix this yet, even tough github list this as outdated now.

**Comment by ChrisTerBeke** [comment]
Yes, just change the version tag on top of the Dockerfile.

**Comment by daid** [comment]
Not seeing a version tag?

**Comment by ChrisTerBeke** [comment]
Then it's using the latest stable.

Change `FROM python:alpine AS base` for `FROM python:3.4-alpine AS base`.

**Comment by ChrisTerBeke** [comment]
Typing is not included in python 3.4 I guess, it needs to be added to requirements.txt.

**Comment by daid** [comment]
And, another typing issue. getStream("/toolpath") returns a IO[bytes] for an UFP and a IO[str] for a GCode file.
I'm fixing it so that the gcode file also returns an IO[bytes].

**Comment by ChrisTerBeke** [comment]
Nice, green CI :)

**Comment by daid** [comment]
Also:
```
d.braam@um-laptop-111 MINGW32 /c/software/Monolith/libCharon (FixTypingErrors)
$ /c/python34/Scripts/mypy Charon/ "--ignore-missing-imports"

d.braam@um-laptop-111 MINGW32 /c/software/Monolith/libCharon (FixTypingErrors)
$
```


But with strict typing:
```
d.braam@um-laptop-111 MINGW32 /c/software/Monolith/libCharon (FixTypingErrors)
$ /c/python34/Scripts/mypy Charon/ "--ignore-missing-imports" --strict
Charon\Service\FileService.py:20: error: Class cannot subclass 'Object' (has type 'Any')
Charon\Service\FileService.py:43: error: Function is missing a type annotation
Charon\Service\FileService.py:58: error: Function is missing a type annotation
Charon\Service\FileService.py:71: error: Function is missing a type annotation
Charon\Service\FileService.py:80: error: Function is missing a type annotation
Charon\Service\FileService.py:88: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:62: error: Function is missing a type annotation for one or more arguments
Charon\Client\DBusInterface.py:63: error: Call to untyped function "__ensureDBusSetup" in typed context
Charon\Client\DBusInterface.py:63: error: Call to untyped function "__ensureDBusSetup" of "DBusInterface" in typed context
Charon\Client\DBusInterface.py:95: error: Function is missing a type annotation for one or more arguments
Charon\Client\DBusInterface.py:96: error: Call to untyped function "__ensureDBusSetup" in typed context
Charon\Client\DBusInterface.py:96: error: Call to untyped function "__ensureDBusSetup" of "DBusInterface" in typed context
Charon\Client\DBusInterface.py:104: error: Call to untyped function "asyncCall" in typed context
Charon\Client\DBusInterface.py:121: error: Call to untyped function "__ensureDBusSetup" in typed context
Charon\Client\DBusInterface.py:121: error: Call to untyped function "__ensureDBusSetup" of "DBusInterface" in typed context
Charon\Client\DBusInterface.py:125: warning: Returning Any from function declared to return "bool"
Charon\Client\DBusInterface.py:125: error: Call to untyped function "addConnection" in typed context
Charon\Client\DBusInterface.py:144: error: Call to untyped function "__ensureDBusSetup" in typed context
Charon\Client\DBusInterface.py:144: error: Call to untyped function "__ensureDBusSetup" of "DBusInterface" in typed context
Charon\Client\DBusInterface.py:148: warning: Returning Any from function declared to return "bool"
Charon\Client\DBusInterface.py:148: error: Call to untyped function "removeConnection" in typed context
Charon\Client\DBusInterface.py:156: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:166: error: Call to untyped function "DBusSignalForwarder" in typed context
Charon\Client\DBusInterface.py:190: error: Class cannot subclass 'QObject' (has type 'Any')
Charon\Client\DBusInterface.py:191: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:196: error: Need type annotation for '__interface_objects'
Charon\Client\DBusInterface.py:197: error: Need type annotation for '__connected_signals'
Charon\Client\DBusInterface.py:198: error: Need type annotation for '__callbacks'
Charon\Client\DBusInterface.py:200: error: Need type annotation for '__pending_async_calls'
Charon\Client\DBusInterface.py:203: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:214: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:232: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:241: error: Function is missing a type annotation
Charon\Client\DBusInterface.py:248: error: Function is missing a type annotation
Charon\Client\Request.py:43: error: Function is missing a type annotation
Charon\Client\Request.py:88: error: Incompatible default for argument "data" (default has type "None", argument has type "Callable[[Request, Dict[str, Any]], None]")
Charon\Client\Request.py:88: error: Incompatible default for argument "completed" (default has type "None", argument has type "Callable[[Request], None]")
Charon\Client\Request.py:88: error: Incompatible default for argument "error" (default has type "None", argument has type "Callable[[Request, str], None]")
Charon\Client\Request.py:97: error: Function is missing a type annotation
Charon\Client\Request.py:114: error: Function is missing a type annotation
Charon\Client\Request.py:118: error: Argument 2 to "callAsync" of "DBusInterface" has incompatible type "None"; expected "Callable[..., None]"
Charon\Client\Request.py:118: error: Argument 3 to "callAsync" of "DBusInterface" has incompatible type "None"; expected "Callable[..., None]"
Charon\Client\Request.py:125: error: Function is missing a type annotation
Charon\Client\Request.py:132: error: Function is missing a return type annotation
Charon\Client\Request.py:137: error: Function is missing a return type annotation
Charon\Client\Request.py:145: error: Function is missing a return type annotation
Charon\Client\Request.py:157: error: Function is missing a return type annotation
Charon\Client\Request.py:171: error: Function is missing a return type annotation
Charon\Client\Request.py:183: error: Function is missing a type annotation
Charon\VirtualFile.py:32: error: Function is missing a type annotation
Charon\VirtualFile.py:35: error: Function is missing a type annotation
Charon\VirtualFile.py:43: error: Function is missing a type annotation
Charon\VirtualFile.py:45: error: Too many arguments for "openStream" of "FileInterface"
Charon\VirtualFile.py:47: error: Function is missing a type annotation
Charon\VirtualFile.py:56: error: Function is missing a type annotation
Charon\VirtualFile.py:64: error: Function is missing a type annotation
Charon\Service\RequestQueue.py:42: error: Function is missing a type annotation
Charon\Service\RequestQueue.py:44: error: Call to untyped function "VirtualFile" in typed context
Charon\Service\RequestQueue.py:64: error: Argument "exc_info" to "log" of "Logger" has incompatible type "int"; expected "Union[None, bool, Tuple[type, BaseException, TracebackType], Tuple[None, None, None]]"
Charon\Service\RequestQueue.py:92: error: Function is missing a type annotation
Charon\Service\RequestQueue.py:93: error: Need type annotation for '__queue'
Charon\Service\RequestQueue.py:98: error: Need type annotation for '__request_map'
Charon\Service\RequestQueue.py:100: error: Need type annotation for '__workers'
Charon\Service\RequestQueue.py:112: error: Function is missing a return type annotation
Charon\Service\RequestQueue.py:118: error: Cannot determine type of '__queue'
Charon\Service\RequestQueue.py:131: error: Function is missing a return type annotation
Charon\Service\RequestQueue.py:145: error: Cannot determine type of '__queue'
Charon\Service\RequestQueue.py:147: warning: Returning Any from function declared to return "Request"
Charon\Service\RequestQueue.py:150: error: Function is missing a type annotation
Charon\Service\RequestQueue.py:159: error: Argument "exc_info" to "log" of "Logger" has incompatible type "int"; expected "Union[None, bool, Tuple[type, BaseException, TracebackType], Tuple[None, None, None]]"
Charon\Client\test_qt.py:14: error: Call to untyped function "start" in typed context
Charon\Client\test_glib.py:20: warning: Returning Any from function declared to return "None"
Charon\Client\test_glib.py:22: error: Call to untyped function "start" in typed context

d.braam@um-laptop-111 MINGW32 /c/software/Monolith/libCharon (FixTypingErrors)
$
```

**Comment by genodeftest** [comment]
Sorry to jump in, but have you found a solution for this error message yet?
> No library stub file for module 'gi.repository'

**Comment by Ghostkeeper** [comment]
Not really. We don't run with strict type checking so it hasn't been a problem for our CI. I'm not even sure what that `gi.repository` is; it's not in our code.

**Comment by genodeftest** [comment]
Thanks for the explaination!
`gi.repository` is the GObject Introspection package used for python access to GLib/Gtk code such as the DBus code in [DBusInterface.py](https://github.com/Ultimaker/libCharon/blob/master/Charon/Client/DBusInterface.py).

**Comment by Ghostkeeper** [comment]
Ah, I see. I've only really used the base of this repo and the filetypes module, not the DBus part.

Maybe it helps to install these stubs: https://pypi.org/project/PyGObject-stubs/

**Comment by genodeftest** [comment]
> Maybe it helps to install these stubs: https://pypi.org/project/PyGObject-stubs/

Or even better: https://github.com/strycore/fakegir and the long term solution at https://gitlab.gnome.org/GNOME/pygobject/issues/159. Thanks!

**Pull request merged**